File: go/socketd-example/demo02_base/listener.go
package demo02_base

import (
	"fmt"
	"log/slog"
	"os"
	"strings"
	"time"

	"socketd/transport/core"
	"socketd/transport/core/message"
)

func SetLog() {
	slog.SetDefault(slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
		//AddSource: true,
		Level: slog.LevelDebug,
		ReplaceAttr: func(groups []string, a slog.Attr) slog.Attr {
			if a.Key == slog.TimeKey {
				if t, ok := a.Value.Any().(time.Time); ok {
					a.Value = slog.StringValue(t.Format(time.DateTime))
				}
			}
			return a
		},
	})))
}

type MyListner struct {
}

func (l *MyListner) OnClose(session core.Session) {
	fmt.Println("-> OnClose")
}
func (l *MyListner) OnError(session core.Session, err error) {
	fmt.Println("-> OnError")
}
func (l *MyListner) OnOpen(session core.Session) error {
	fmt.Println("-> OnOpen")
	return nil
}
func (l *MyListner) OnMessage(session core.Session, msg *message.Frame) error {
	fmt.Println("-> OnMessage")
	if strings.HasPrefix(msg.Message.Event, "/demo") {

		ent := message.NewEntity(nil, []byte("Nice to meet U :)"))

		err := session.Reply(msg, ent)
		if err != nil {
			fmt.Println(err)
			return err
		}
	}
	return nil
}

File: java/socketd/src/main/java/org/noear/socketd/transport/core/Listener.java
package org.noear.socketd.transport.core;

import java.io.IOException;

/**
 * 监听器
 *
 * @author noear
 * @since 2.0
 */
public interface Listener {
    /**
     * 打开时
     *
     * @param session 会话
     */
    void onOpen(Session session) throws IOException;

    /**
     * 收到消息时
     *
     * @param session 会话
     * @param message 消息
     */
    void onMessage(Session session, Message message) throws IOException;

    /**
     * 收到答复时
     *
     * @param session 会话
     * @param message 消息
     */
    default void onReply(Session session, Message message) {
    }

    /**
     * 发送消息时
     *
     * @param session 会话
     * @param message 消息
     */
    default void onSend(Session session, Message message) {
    }

    /**
     * 关闭时
     *
     * @param session 会话
     */
    void onClose(Session session);

    /**
     * 出错时
     *
     * @param session 会话
     * @param error   错误信息
     */
    void onError(Session session, Throwable error);
}